src/Microsoft.AspNetCore.Razor.Tasks/DotnetToolTask.cs

```diff
@@ -82,10 +103,102 @@ protected override void LogToolCommand(string message)
             }
         }
 
-        protected virtual bool TryExecuteOnServer(out bool result)
+        public override void Cancel()
         {
-            result = false;
+            base.Cancel();
+
+            _razorServerCts?.Cancel();
+        }
+
+        protected virtual bool TryExecuteOnServer(string pathToTool, string responseFileCommands, string commandLineCommands, out int result)
+        {
+            CompilerServerLogger.Log("Server execution started.");
+            using (_razorServerCts = new CancellationTokenSource())
+            {
+                CompilerServerLogger.Log($"CommandLine = '{commandLineCommands}'");
+                CompilerServerLogger.Log($"BuildResponseFile = '{responseFileCommands}'");
+
+                var clientDir = Path.GetDirectoryName(ServerAssembly);
+
+                var workingDir = CurrentDirectoryToUse();
+                var buildPaths = new BuildPathsAlt(
+                    clientDir: clientDir,
+                    // MSBuild doesn't need the .NET SDK directory
+                    sdkDir: null,
+                    workingDir: workingDir,
+                    tempDir: BuildServerConnection.GetTempPath(workingDir));
+
+                var responseTask = BuildServerConnection.RunServerCompilation(
+                    Command,
+                    GetArguments(responseFileCommands),
+                    buildPaths,
+                    keepAlive: null,
+                    libEnvVariable: LibDirectoryToUse(),
+                    cancellationToken: _razorServerCts.Token);
+
+                responseTask.Wait(_razorServerCts.Token);
+
+                var response = responseTask.Result;
+                if (response.Type == BuildResponse.ResponseType.Completed &&
+                    response is CompletedBuildResponse completedResponse)
+                {
+                    CompilerServerLogger.Log("Server execution completed.");
+
+                    result = completedResponse.ReturnCode;
```

Comment:
if result is non-zero her we may want to log something. Otherwise we'll be failing to invoke the server without an error.